src/link/script.c

```diff
@@ -0,0 +1,214 @@
+		if (bank >= BANK_COUNT_ROMX)
+			errx(1, "ROMX index too big (%d >= %d).\n", bank, BANK_COUNT_ROMX);
+		current_bank = BANK_ROMX + bank - 1;
+		current_real_bank = bank;
+		return;
+	} else if (strcmp(type, "VRAM") == 0) {
+		if (bank >= BANK_COUNT_VRAM)
+			errx(1, "VRAM index too big (%d >= %d).\n", bank, BANK_COUNT_VRAM);
+		current_bank = BANK_VRAM + bank;
+		current_real_bank = bank;
+		return;
+	} else if (strcmp(type, "WRAM0") == 0) {
+		if (bank != 0)
+			errx(1, "(Internal) Trying to assign a bank number to WRAM0.\n");
+		current_bank = BANK_WRAM0;
+		current_real_bank = 0;
+		return;
+	} else if (strcmp(type, "WRAMX") == 0) {
+		if (bank >= BANK_COUNT_WRAMX)
+			errx(1, "WRAMX index too big (%d >= %d).\n", bank, BANK_COUNT_WRAMX);
+		current_bank = BANK_WRAMX + bank - 1;
+		current_real_bank = bank - 1;
+		return;
+	} else if (strcmp(type, "SRAM") == 0) {
+		if (bank >= BANK_COUNT_SRAM)
+			errx(1, "SRAM index too big (%d >= %d).\n", bank, BANK_COUNT_SRAM);
+		current_bank = BANK_SRAM + bank;
+		current_real_bank = bank;
+		return;
+	} else if (strcmp(type, "OAM") == 0) {
+		if (bank != 0)
+			errx(1, "(Internal) Trying to assign a bank number to OAM.\n");
+		current_bank = BANK_OAM;
+		current_real_bank = 0;
+		return;
+	} else if (strcmp(type, "HRAM") == 0) {
+		if (bank != 0)
+			errx(1, "(Internal) Trying to assign a bank number to HRAM.\n");
+		current_bank = BANK_HRAM;
+		current_real_bank = 0;
+		return;
+	}
+
+	errx(1, "(Internal) Unknown section type [%s].\n", type);
+}
+
+void script_SetAddress(unsigned int addr)
+{
+	if (current_bank == -1) {
+		errx(1, "Trying to set an address without assigned bank\n");
+	}
+
+	/* Make sure that we don't go back. */
+	if (bank[current_bank].address > addr) {
+		errx(1, "Trying to go to a previous address (0x%04X to 0x%04X)\n",
+			bank[current_bank].address, addr);
+	}
+
+	bank[current_bank].address = addr;
+
+	/* Make sure we don't overflow */
+	if (bank[current_bank].address >= bank[current_bank].top_address) {
+		errx(1, "Bank overflowed (0x%04X >= 0x%04X)\n",
+			bank[current_bank].address, bank[current_bank].top_address);
+	}
+}
+
+void script_SetAlignment(unsigned int alignment)
+{
+	if (current_bank == -1) {
+		errx(1, "Trying to set an alignment without assigned bank\n");
+	}
+
+	if (alignment > 15) {
+		errx(1, "Trying to set an alignment too big: %d\n", alignment);
+	}
+
+	unsigned int size = 1 << alignment;
+	unsigned int mask = size - 1;
+
+	if (bank[current_bank].address & mask) {
+		bank[current_bank].address &= ~mask;
+		bank[current_bank].address += size;
+	}
+
+	/* Make sure we don't overflow */
+	if (bank[current_bank].address >= bank[current_bank].top_address) {
+		errx(1, "Bank overflowed (0x%04X >= 0x%04X)\n",
+			bank[current_bank].address, bank[current_bank].top_address);
+	}
+}
+
+void script_OutputSection(const char *section_name)
+{
+	if (current_bank == -1) {
+		errx(1, "Trying to place a section without assigned bank\n");
```

Comment:
More thoughts for the future: It might not be infeasible to ask that sections come after each other, but we don't care which bank they're in. Not really a priority, though.